cranelift/codegen/src/opts/algebraic.isle

```diff
@@ -198,6 +196,30 @@
       (let ((mask Imm64 (imm64_shl ty (imm64 0xFFFF_FFFF_FFFF_FFFF) k)))
         (band ty x (iconst ty mask))))
 
+;; For unsigned shifts, `(x << k) >> k` is the same as masking out the top
+;; `k` bits. A similar rule is valid for vectors but this `iconst` mask only
+;; works for scalar integers.
+(rule (simplify (ushr (fits_in_64 (ty_int ty))
+                      (ishl ty x (iconst _ k))
+                      (iconst _ k)))
+      (band ty x (iconst ty (imm64_ushr ty (imm64 (ty_mask ty)) k))))
```

Comment:
I'm a little confused about the requirements for `iconst`. Its documentation says it can create "an integer vector where all the lanes have the same value", which probably should work here. Do I just pass the lane type instead of the original `ty` in some of these places? `ty_mask`, at least, and also maybe `imm64_ushr` I guess?

Reply:
Hmm -- that feature is somewhat surprising to me! I don't think we actually have lowerings that do that (and in fact I think we have another opcode for "vector broadcast" or whatever the proper name for that operation is)?

Reply:
> and in fact I think we have another opcode for "vector broadcast" or whatever the proper name for that operation is

It's called `splat`.